File: python/sonara/vocal_model.py
"""Bring-your-own vocalness model: a pure-numpy trainer + loader for a tiny
vocal-presence classifier over sonara's similarity embedding.

The built-in ``vocalness`` heuristic cannot separate clean solo voice from
clean solo pitched instruments. This module trains the precision upgrade and
hands it back to ``analyze_*`` via ``vocalness_model=<path>``:

    import sonara
    from sonara import vocal_model

    # 1. Embeddings for a labeled set (True = vocal, False = instrumental).
    lib = sonara.analyze_batch(files, features=["embedding"])
    X = [r["embedding"] for r in lib]      # (n, 48)
    y = [True, False, ...]                 # vocal presence per track

    # 2. Train (numpy only) and save. `model_id` is REQUIRED — it is stamped
    #    into every result's provenance as `vocalness_model_id` so downstream
    #    caches can invalidate scores produced by a different model.
    model = vocal_model.train(X, y, model_id="my-vocal-v1", hidden=16)
    model.save("vocal_model.json")

    # 3. Score new tracks: vocalness/instrumentalness now come from the model.
    r = sonara.analyze_file("track.mp3", vocalness_model="vocal_model.json")
    r["vocalness"], r["provenance"]["vocalness_model_id"]

The JSON is the genre-model format (last layer softmax over exactly two
labels, one of them ``"vocal"``) plus the required ``id``. Training
standardizes features internally and folds the standardization into the first
layer's weights, so the saved model operates on raw embeddings — identical
inference in Rust and numpy.
"""

from __future__ import annotations

import json

import numpy as np

from sonara._sonara import SIMILARITY_VERSION as _SIMILARITY_VERSION
from sonara._sonara import EMBEDDING_DIM as _EMBEDDING_DIM

FORMAT_VERSION = 1
LABELS = ["instrumental", "vocal"]  # fixed; index 1 is P(vocal)

__all__ = ["VocalnessModel", "train", "load", "bundled_path", "FORMAT_VERSION",
           "LABELS"]


def bundled_path() -> str:
    """Path to the vocalness model bundled with the package.

    Trained on ~900 album-level labeled tracks; on a held-out 205-track
    curated real-music set: AUC 0.94 vs 0.63 for the built-in heuristic,
    clear-vocal false negatives 5% vs 53%. Pass it (or the shorthand
    ``vocalness_model="bundled"``) to ``analyze_*``. Known limitations:
    solo melodic instrument leads (sax/guitar) can score vocal-high, and
    spoken narration scores low.
    """
    import os

    return os.path.join(os.path.dirname(__file__), "models", "vocalness_v1.json")


def _softmax_rows(m: np.ndarray) -> np.ndarray:
    z = m - m.max(axis=1, keepdims=True)
    e = np.exp(z)
    return e / e.sum(axis=1, keepdims=True)


class VocalnessModel:
    """A trained (or loaded) vocal-presence classifier over the embedding."""

    def __init__(self, model_id, layers, embedding_version):
        self.id = str(model_id)
        if not self.id.strip():
            raise ValueError("model_id must be a non-empty string")
        self.layers = layers  # [{"W": (in, out), "b": (out,), "activation": str}]
        self.embedding_version = int(embedding_version)

    def predict_vocalness(self, x) -> float:
        """P(vocal) for one embedding vector, in ``[0, 1]`` (numpy parity
        with the Rust ``VocalnessModel::predict_vocalness``)."""
        v = np.asarray(x, dtype=np.float64).ravel()
        v = np.where(np.isfinite(v), v, 0.0)
        for layer in self.layers:
            v = v @ layer["W"] + layer["b"]
            act = layer["activation"]
            if act == "relu":
                v = np.maximum(v, 0.0)
            elif act == "softmax":
                z = v - np.max(v)
                e = np.exp(z)
                v = e / e.sum()
        return float(np.clip(v[1], 0.0, 1.0))

    def to_dict(self):
        return {
            "format_version": FORMAT_VERSION,
            "embedding_version": self.embedding_version,
            "id": self.id,
            "labels": list(LABELS),
            "layers": [
                {
                    # JSON is row-major out x in — transpose the internal (in, out).
                    "weights": np.asarray(layer["W"]).T.tolist(),
                    "bias": np.asarray(layer["b"]).ravel().tolist(),
                    "activation": layer["activation"],
                }
                for layer in self.layers
            ],
        }

    def save(self, path):
        """Write the model as JSON to ``path`` (loadable by ``analyze_*``)."""
        with open(path, "w", encoding="utf-8") as f:
            json.dump(self.to_dict(), f)
        return path


def load(path) -> VocalnessModel:
    """Load a vocalness model JSON written by :meth:`VocalnessModel.save`."""
    with open(path, "r", encoding="utf-8") as f:
        d = json.load(f)
    labels = [str(x).lower() for x in d["labels"]]
    if labels != LABELS:
        raise ValueError(f"labels must be {LABELS}; got {d['labels']}")
    layers = []
    for layer in d["layers"]:
        w = np.asarray(layer["weights"], dtype=np.float64)
        layers.append(
            {
                "W": w.T.copy(),
                "b": np.asarray(layer["bias"], dtype=np.float64),
                "activation": str(layer["activation"]),
            }
        )
    return VocalnessModel(d["id"], layers, d["embedding_version"])


def train(X, y, *, model_id, hidden=16, epochs=2000, lr=0.05, seed=0, l2=1e-3,
          embedding_version=None) -> VocalnessModel:
    """Train a vocal-presence classifier with plain numpy (no sklearn/torch).

    Parameters
    ----------
    X : array-like, shape ``(n, EMBEDDING_DIM)``
        One similarity embedding per training example.
    y : sequence of bool (or 0/1), length ``n``
        ``True``/1 = vocal, ``False``/0 = instrumental.
    model_id : str (required)
        Model identity stamped into analysis provenance
        (``vocalness_model_id``). Version it like an artifact
        (e.g. ``"myset-vocal-v3"``).
    hidden : int, default 16
        ``0`` → logistic regression; ``> 0`` → one hidden ReLU layer.
    epochs, lr, l2, seed : training hyperparameters (deterministic given seed).
    embedding_version : int, optional
        Recorded in the model; defaults to ``sonara.SIMILARITY_VERSION``.

    Features are standardized (zero mean, unit variance over the training
    set) internally; the standardization is folded into the first layer's
    weights so the saved model operates on raw embeddings.
    """
    X = np.asarray(X, dtype=np.float64)
    if X.ndim != 2 or X.shape[1] != _EMBEDDING_DIM:
        raise ValueError(f"X must have shape (n, {_EMBEDDING_DIM}); got {X.shape}")
    X = np.where(np.isfinite(X), X, 0.0)
    yi = np.array([1 if bool(v) else 0 for v in y], dtype=np.int64)
    if len(yi) != X.shape[0]:
        raise ValueError(f"len(y) ({len(yi)}) must equal X.shape[0] ({X.shape[0]})")
    if len(set(yi.tolist())) < 2:
        raise ValueError("need both vocal and instrumental examples")

    # Standardize (folded into the first layer below).
    mu = X.mean(axis=0)
    sigma = X.std(axis=0)
    sigma[sigma < 1e-8] = 1.0
    Xs = (X - mu) / sigma

    n, d = X.shape
    k = 2
    Y = np.zeros((n, k), dtype=np.float64)
    Y[np.arange(n), yi] = 1.0
    # Class weighting: balance vocal/instrumental contributions.
    counts = np.bincount(yi, minlength=2).astype(np.float64)
    w_sample = (n / (2.0 * counts))[yi][:, None]

    rng = np.random.default_rng(seed)
    ev = _SIMILARITY_VERSION if embedding_version is None else int(embedding_version)

    if hidden and hidden > 0:
        h = int(hidden)
        W1 = rng.normal(0.0, np.sqrt(2.0 / d), size=(d, h))
        b1 = np.zeros(h)
        W2 = rng.normal(0.0, np.sqrt(2.0 / h), size=(h, k))
        b2 = np.zeros(k)
        for _ in range(int(epochs)):
            Z1 = Xs @ W1 + b1
            A1 = np.maximum(Z1, 0.0)
            P = _softmax_rows(A1 @ W2 + b2)
            dZ2 = (P - Y) * w_sample / n
            dW2 = A1.T @ dZ2 + l2 * W2
            db2 = dZ2.sum(axis=0)
            dZ1 = (dZ2 @ W2.T) * (Z1 > 0.0)
            dW1 = Xs.T @ dZ1 + l2 * W1
            db1 = dZ1.sum(axis=0)
            W1 -= lr * dW1
            b1 -= lr * db1
            W2 -= lr * dW2
            b2 -= lr * db2
        # Fold standardization into layer 1: x_s = (x - mu) / sigma.
        W1_raw = W1 / sigma[:, None]
        b1_raw = b1 - (mu / sigma) @ W1
        layers = [
            {"W": W1_raw, "b": b1_raw, "activation": "relu"},
            {"W": W2, "b": b2, "activation": "softmax"},
        ]
    else:
        W = rng.normal(0.0, 0.1, size=(d, k))
        b = np.zeros(k)
        for _ in range(int(epochs)):
            P = _softmax_rows(Xs @ W + b)
            dZ = (P - Y) * w_sample / n
            dW = Xs.T @ dZ + l2 * W
            db = dZ.sum(axis=0)
            W -= lr * dW
            b -= lr * db
        W_raw = W / sigma[:, None]
        b_raw = b - (mu / sigma) @ W
        layers = [{"W": W_raw, "b": b_raw, "activation": "softmax"}]

    return VocalnessModel(model_id, layers, ev)

File: sonara/src/vocal_model.rs
//! Bring-your-own vocal-presence classification: a small feed-forward model
//! over the hand-crafted similarity embedding, replacing the built-in
//! `vocalness` heuristic when supplied.
//!
//! The built-in contrast-based heuristic (see `TrackAnalysis::vocalness`)
//! cannot separate clean solo voice from clean solo pitched instruments —
//! on a labeled real-music set it scores AUC ≈ 0.63 with a large clear-vocal
//! false-negative cluster. This module is the precision upgrade socket: a
//! trained classifier (same JSON MLP format as [`crate::genre`], trainer in
//! `python/sonara/vocal_model.py`) that overrides `vocalness` and
//! `instrumentalness` when set via
//! [`AnalysisConfig::vocalness_model`](crate::analyze::AnalysisConfig).
//!
//! ## Format
//!
//! The JSON layout is exactly the genre-model format
//! ([`crate::genre::GENRE_MODEL_FORMAT_VERSION`]) with three extra
//! requirements, validated at load:
//! - `id` is **required** (non-empty): the model identity is carried into
//!   [`AnalysisProvenance`](crate::analyze::AnalysisProvenance) as
//!   `vocalness_model_id`, so downstream caches can invalidate scores
//!   produced by a different (or no) model.
//! - exactly **two labels**, one of which must be `"vocal"`
//!   (case-insensitive); the other conventionally `"instrumental"`.
//! - the usual `embedding_version` match against
//!   [`crate::similarity::SIMILARITY_VERSION`] is enforced at use time.
//!
//! The reported score is the softmax probability of the `"vocal"` label —
//! a calibrated `[0, 1]` vocal-presence estimate, not an argmax.

use std::path::Path;

use crate::error::{Result, SonaraError};
use crate::genre::GenreModel;
use crate::types::Float;

/// A loaded, validated vocal-presence classifier over the similarity
/// embedding. Wraps the shared MLP machinery from [`crate::genre`].
#[derive(Debug, Clone)]
pub struct VocalnessModel {
    inner: GenreModel,
    /// Index of the `"vocal"` label in the softmax output.
    vocal_idx: usize,
}

impl VocalnessModel {
    /// The model identity string (required at load, never empty).
    pub fn id(&self) -> &str {
        self.inner.id.as_deref().unwrap_or("")
    }

    /// The `embedding_version` the model was trained against.
    pub fn embedding_version(&self) -> u32 {
        self.inner.embedding_version
    }

    /// P(vocal) for an embedding vector, in `[0, 1]`.
    pub fn predict_vocalness(&self, embedding: &[Float]) -> Float {
        self.try_predict_vocalness(embedding).unwrap_or(0.0)
    }

    /// Fallible P(vocal) inference for analysis paths that must surface
    /// malformed model state rather than emitting a non-finite value.
    pub fn try_predict_vocalness(&self, embedding: &[Float]) -> Result<Float> {
        let probs = self.inner.try_predict_probs(embedding)?;
        let probability = probs
            .get(self.vocal_idx)
            .copied()
            .ok_or_else(|| SonaraError::ModelError("vocalness output is missing".into()))?;
        if !probability.is_finite() || !(0.0..=1.0).contains(&probability) {
            return Err(SonaraError::ModelError(
                "vocalness probability must be finite and in [0,1]".into(),
            ));
        }
        Ok(probability)
    }

    fn validate(inner: GenreModel) -> Result<Self> {
        let err = |msg: &str| SonaraError::ModelError(format!("vocalness model: {msg}"));
        match inner.id.as_deref() {
            Some(id) if !id.trim().is_empty() => {}
            _ => {
                return Err(err(
                    "`id` is required (non-empty) — it identifies the model in analysis provenance",
                ))
            }
        }
        if inner.labels.len() != 2 {
            return Err(err("exactly two labels required (vocal + instrumental)"));
        }
        let vocal_idx = inner
            .labels
            .iter()
            .position(|l| l.eq_ignore_ascii_case("vocal"))
            .ok_or_else(|| err("one label must be \"vocal\""))?;
        Ok(Self { inner, vocal_idx })
    }
}

/// The vocalness model bundled with the crate, as JSON
/// (`models/vocalness_v2.json`, embedded at compile time). Identical to the
/// artifact the Python package resolves for `vocalness_model="bundled"`.
const BUNDLED_JSON: &str = include_str!("../models/vocalness_v2.json");

/// Load the vocalness model bundled with the crate (`sonara-vocalness-v2`).
///
/// This is the Rust-native equivalent of the Python
/// `vocalness_model="bundled"` shorthand: the ~33 KB JSON artifact is embedded
/// in the crate at compile time (dead-stripped when unused), so in-process
/// consumers of the Rust core need no filesystem lookup or vendored copy.
/// Validation numbers and known limitations are documented on the release.
///
/// Fallible by signature for API stability across future model revisions;
/// with the shipped artifact it always succeeds (covered by tests).
pub fn bundled() -> Result<VocalnessModel> {
    from_json_str(BUNDLED_JSON)
}

/// Parse and validate a vocalness model from a JSON string.
pub fn from_json_str(s: &str) -> Result<VocalnessModel> {
    VocalnessModel::validate(crate::genre::from_json_str(s)?)
}

/// Load and validate a vocalness model from a JSON file on disk.
pub fn load(path: &Path) -> Result<VocalnessModel> {
    VocalnessModel::validate(crate::genre::load(path)?)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn model_json(id: &str, labels: &str) -> String {
        // 48 → 2 single-layer softmax; weights favor the second label a bit.
        let row0: Vec<String> = (0..48).map(|_| "0.0".to_string()).collect();
        let row1: Vec<String> = (0..48).map(|i| format!("{}", (i as f32) * 0.01)).collect();
        format!(
            r#"{{"format_version":1,"embedding_version":{ev},"id":{id},"labels":{labels},
                 "layers":[{{"weights":[[{r0}],[{r1}]],"bias":[0.0,0.1],"activation":"softmax"}}]}}"#,
            ev = crate::similarity::SIMILARITY_VERSION,
            r0 = row0.join(","),
            r1 = row1.join(","),
        )
    }

    #[test]
    fn test_valid_model_loads_and_predicts() {
        let m = from_json_str(&model_json(
            "\"vocal-test-v1\"",
            r#"["instrumental","vocal"]"#,
        ))
        .unwrap();
        assert_eq!(m.id(), "vocal-test-v1");
        let p_zero = m.predict_vocalness(&[0.0; 48]);
        assert!((0.0..=1.0).contains(&p_zero));
        let p_ones = m.predict_vocalness(&[1.0; 48]);
        assert!(p_ones > p_zero, "positive weights on vocal row must raise P");
        // Deterministic.
        assert_eq!(p_ones, m.predict_vocalness(&[1.0; 48]));
    }

    #[test]
    fn test_missing_id_rejected() {
        let json = model_json("null", r#"["instrumental","vocal"]"#);
        assert!(from_json_str(&json).is_err());
        let json = model_json("\"  \"", r#"["instrumental","vocal"]"#);
        assert!(from_json_str(&json).is_err());
    }

    #[test]
    fn test_label_shape_rejected() {
        // Three labels.
        let json = model_json("\"m\"", r#"["a","vocal","c"]"#);
        assert!(from_json_str(&json).is_err());
        // No "vocal" label. (Two labels, but neither is vocal.)
        let row: Vec<String> = (0..48).map(|_| "0.0".into()).collect();
        let json = format!(
            r#"{{"format_version":1,"embedding_version":{},"id":"m","labels":["a","b"],
                 "layers":[{{"weights":[[{r}],[{r}]],"bias":[0.0,0.0],"activation":"softmax"}}]}}"#,
            crate::similarity::SIMILARITY_VERSION,
            r = row.join(","),
        );
        assert!(from_json_str(&json).is_err());
    }

    #[test]
    fn test_vocal_label_position_respected() {
        // "vocal" first: P(vocal) must read index 0.
        let m = from_json_str(&model_json(
            "\"vocal-first\"",
            r#"["vocal","instrumental"]"#,
        ))
        .unwrap();
        // Row 1 (instrumental here) has the positive weights, so ones input
        // should LOWER P(vocal).
        let p_zero = m.predict_vocalness(&[0.0; 48]);
        let p_ones = m.predict_vocalness(&[1.0; 48]);
        assert!(p_ones < p_zero);
    }

    #[test]
    fn test_bundled_loads_with_stable_id() {
        let m = bundled().expect("bundled model must always load");
        assert_eq!(m.id(), "sonara-vocalness-v2");
        assert_eq!(m.embedding_version(), crate::similarity::SIMILARITY_VERSION);
        let p = m.predict_vocalness(&[0.0; 48]);
        assert!(p.is_finite() && (0.0..=1.0).contains(&p));
    }

    #[test]
    fn test_bundled_matches_python_package_artifact() {
        // The Python wheel ships its own copy as package data; the two must
        // stay byte-identical. Runs only in the repo layout (both files
        // present) — exactly where a divergence could be introduced.
        let py_copy = concat!(
            env!("CARGO_MANIFEST_DIR"),
            "/../python/sonara/models/vocalness_v2.json"
        );
        match std::fs::read_to_string(py_copy) {
            Ok(py_json) => assert_eq!(
                py_json, BUNDLED_JSON,
                "sonara/models/vocalness_v2.json and \
                 python/sonara/models/vocalness_v2.json have diverged"
            ),
            Err(_) => {} // vendored-crate layout: python/ not present
        }
    }

    #[test]
    fn test_non_finite_embedding_is_safe() {
        let m = from_json_str(&model_json(
            "\"vocal-test-v1\"",
            r#"["instrumental","vocal"]"#,
        ))
        .unwrap();
        let mut emb = [1.0 as Float; 48];
        emb[3] = Float::NAN;
        emb[7] = Float::INFINITY;
        let p = m.predict_vocalness(&emb);
        assert!(p.is_finite() && (0.0..=1.0).contains(&p));
    }

    #[test]
    fn test_fallible_prediction_rejects_non_finite_model_output() {
        let mut model = from_json_str(&model_json(
            "\"vocal-test-v1\"",
            r#"["instrumental","vocal"]"#,
        ))
        .unwrap();
        let layer = model.inner.layers.last_mut().unwrap();
        layer.bias[0] = Float::MAX;
        layer.weights[0][0] = Float::MAX;

        let embedding = [2.0; 48];
        assert!(matches!(
            model.try_predict_vocalness(&embedding),
            Err(SonaraError::ModelError(_))
        ));
        assert!(model.predict_vocalness(&embedding).is_finite());
    }
}
